r/R/dataset-format.R

```diff
@@ -139,6 +139,13 @@ FileWriteOptions <- R6Class("FileWriteOptions", inherit = ArrowObject,
         dataset___ParquetFileWriteOptions__update(self,
             ParquetWriterProperties$create(...),
             ParquetArrowWriterProperties$create(...))
+      } else if (self$type == "ipc") {
+        args <- list(...)
+        dataset___IpcFileWriteOptions__update(self,
+            get_ipc_use_legacy_format(args$use_legacy_format),
+            args$codec,
```

Comment:
I think a better approach for `codec` would be to allow you to specify the parameters like `write_feather` (though fine if you want to provide your own Codec I guess). so something like this:

```suggestion
            args$codec %||% Codec$create(
              args$compression %||% "uncompressed", 
              args$compression_level %||% NA
            ),
```

Perhaps this lets you get rid of your funky `nil_value` helper too? With this change, you're always providing a Codec to C++, never `NULL`.

Reply:
I'll add support for `compression, compression_level` but unfortunately `Codec$create("uncompressed")` doesn't work and we'd still need to fall back to either `nil_value` or overloads of `dataset___IpcFileWriteOptions__update`. It would also be my preference to have a non-nullptr codec in all cases but unfortunately the C++ library uses `nullptr` to indicate a lack of compression (there is no "uncompressed" codec), fixing of which is outside scope for this PR.

Reply:
so "uncompressed" codec is a nullptr? that means you can have `std::shared_ptr<arrow::Codec>` in the `dataset___IpcFileWriteOptions__update` signature, and then check for nullptr (and not R NULL)?

Reply:
Unfortunately, the `shared_ptr` utility already does the reverse mapping (c++ `shared_ptr<Codec>(nullptr)` -> R `NULL`) https://github.com/apache/arrow/blob/20cd08a8143eed7b48686ac3e696252dc5fdee88/r/R/arrow-package.R#L118-L124 so:

```r
uncompressed_type <- arrow:::compression_from_name("uncompressed")
uncompressed_codec <- arrow:::shared_ptr(Codec, arrow:::util___Codec__Create(uncompressed_type, NA))
assert_that(is.null(uncompressed_codec))
```

Reply:
Thus we'd either need to alter `arrow:::shared_ptr()` to support nullptrs or we could implement the reverse mapping and allow R `NULL` -> c++ `shared_ptr<T>(nullptr)`. For now, I'll revert `nil_value` and add overloads of `dataset___IpcFileWriteOptions__update`

Reply:
Ok. Given #8256 in flight I don't think we want to mess with `shared_ptr()`, so do what you think is best.